// lavashark/dist/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VERSION = exports.Track = exports.RepeatMode = exports.Player = exports.NodeState = exports.Node = exports.LavaShark = exports.Filters = exports.ConnectionState = void 0;
const LavaShark_1 = require("./lib/LavaShark");
Object.defineProperty(exports, "LavaShark", { enumerable: true, get: function () { return LavaShark_1.LavaShark; } });
const Node_1 = __importStar(require("./lib/Node"));
exports.Node = Node_1.default;
Object.defineProperty(exports, "NodeState", { enumerable: true, get: function () { return Node_1.NodeState; } });
const Player_1 = __importStar(require("./lib/Player"));
exports.Player = Player_1.default;
Object.defineProperty(exports, "ConnectionState", { enumerable: true, get: function () { return Player_1.ConnectionState; } });
Object.defineProperty(exports, "RepeatMode", { enumerable: true, get: function () { return Player_1.RepeatMode; } });
const Track_1 = __importDefault(require("./lib/queue/Track"));
exports.Track = Track_1.default;
const Filters_1 = __importDefault(require("./lib/Filters"));
exports.Filters = Filters_1.default;
const package_json_1 = require("../package.json");
const VERSION = package_json_1.version;
exports.VERSION = VERSION;
