// lavashark/dist/src/lib/queue/UnresolvedTrack.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const formatTime_1 = require("../utils/formatTime");
class UnresolvedTrack {
    title;
    author;
    duration;
    uri;
    source;
    requester;
    lavashark;
    constructor(lavashark, title, author, duration, uri, source, isrc) {
        this.lavashark = lavashark;
        this.title = title;
        this.author = author;
        this.duration = {
            label: (0, formatTime_1.formatTime)(duration ?? 0),
            value: duration ?? 0
        };
        this.uri = uri ?? '';
        this.source = source ?? 'Unknown';
        if (isrc && lavashark.useISRC)
            this.isrc = isrc;
        this.requester = null;
    }
    get query() {
        return this.isrc ? `"${this.isrc}"` : `${this.author} - ${this.title}`;
    }
    async build() {
        let res = await this.lavashark.search(this.query, this.lavashark.unresolvedSearchSource);
        if (res.loadType !== 'SEARCH_RESULT' || !res.tracks.length) {
            if (!this.isrc)
                throw new Error(`Failed to resolve track ${this.uri}`);
            res = await this.lavashark.search(`${this.author} - ${this.title}`, this.lavashark.unresolvedSearchSource);
            if (res.loadType !== 'SEARCH_RESULT' || !res.tracks.length)
                throw new Error(`Failed to resolve track ${this.uri}`);
        }
        const track = res.tracks[0];
        track.setRequester(this.requester);
        track.metadata = {
            title: this.title,
            author: this.author,
            duration: this.duration,
            uri: this.uri,
            source: this.source,
            isrc: this.isrc,
        };
        return track;
    }
    setRequester(requester) {
        this.requester = requester;
    }
}
exports.default = UnresolvedTrack;
