// lavashark/dist/src/lib/rest/RESTPaths.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.STATS = exports.INFO = exports.VERSION = exports.ROUTE_PLANNER_FREE_ALL = exports.ROUTE_PLANNER_FREE_ADDR = exports.ROUTE_PLANNER_STATUS = exports.DECODE_TRACKS = exports.LOAD_TRACKS = exports.PLAYER = exports.SESSIONS = exports.API_VERSION = void 0;
// API version
exports.API_VERSION = 3;
// Sessions
const SESSIONS = (sessionId) => `/v${exports.API_VERSION}/sessions/${sessionId}`;
exports.SESSIONS = SESSIONS;
// Player
const PLAYER = (sessionId, guildId) => `/v${exports.API_VERSION}/sessions/${sessionId}/players/${guildId}`;
exports.PLAYER = PLAYER;
const LOAD_TRACKS = (identifier) => `/v${exports.API_VERSION}/loadtracks?identifier=${encodeURIComponent(identifier)}`;
exports.LOAD_TRACKS = LOAD_TRACKS;
const DECODE_TRACKS = () => `/v${exports.API_VERSION}/decodetracks`;
exports.DECODE_TRACKS = DECODE_TRACKS;
// Route planner
const ROUTE_PLANNER_STATUS = () => `/v${exports.API_VERSION}/routeplanner/status`;
exports.ROUTE_PLANNER_STATUS = ROUTE_PLANNER_STATUS;
const ROUTE_PLANNER_FREE_ADDR = () => `/v${exports.API_VERSION}/routeplanner/free`;
exports.ROUTE_PLANNER_FREE_ADDR = ROUTE_PLANNER_FREE_ADDR;
const ROUTE_PLANNER_FREE_ALL = () => `/v${exports.API_VERSION}/routeplanner/free/all`;
exports.ROUTE_PLANNER_FREE_ALL = ROUTE_PLANNER_FREE_ALL;
// Server info
const VERSION = () => '/version';
exports.VERSION = VERSION;
const INFO = () => `/v${exports.API_VERSION}/info`;
exports.INFO = INFO;
const STATS = () => `/v${exports.API_VERSION}/stats`;
exports.STATS = STATS;
